clippy_lints/src/matches/redundant_guard.rs

```diff
@@ -0,0 +1,244 @@
+            if let ExprKind::Path(qpath) = expr.kind
+                && let Res::Local(local_ref) = cx.qpath_res(&qpath, expr.hir_id)
+                && *local == local_ref
+            {
+                return ControlFlow::Break(());
+            }
+
+            ControlFlow::Continue(())
+        })
+        .is_none()
+    {
+        return true;
+    }
+
+    false
+}
+
+#[expect(clippy::too_many_arguments)]
+fn emit_redundant_guard(
+    cx: &LateContext<'_>,
+    outer_arm: &Arm<'_>,
+    guard_span: Span,
+    pat_binding: Span,
+    field_name: Option<Ident>,
+    pat_span: Span,
+    inner_guard: Option<Guard<'_>>,
+    app: &mut Applicability,
+) {
+    let (pat_start, pat_end) = remove_binding(cx, outer_arm.pat, pat_binding, app);
+    let binding_replacement = snippet_with_applicability(cx, pat_span, "<binding_repl>", app);
+
+    span_lint_and_then(
+        cx,
+        REDUNDANT_GUARD,
+        guard_span.source_callsite(),
+        "redundant guard",
+        |diag| {
+            diag.span_suggestion(
+                outer_arm.span.with_hi(guard_span.source_callsite().hi()),
+                "try",
+                format!(
+                    "{pat_start}{}{binding_replacement}{pat_end}{}",
+                    field_name.map_or_else(|| Cow::Borrowed(""), |i| format!("{}: ", i.as_str()).into()),
+                    get_guard(cx, inner_guard, app),
+                ),
+                *app,
+            );
+        },
+    );
+}
+
+/// Checks if the given `Expr` can also be represented as a `Pat`.
+fn expr_can_be_pat(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+    fn helper(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+        for_each_expr(expr, |expr| {
+            let expr = expr.peel_blocks();
+            if match expr.kind {
+                ExprKind::ConstBlock(..) => cx.tcx.features().inline_const_pat,
+                ExprKind::Call(c, ..) if let ExprKind::Path(qpath) = c.kind => {
+                    // Allow ctors
+                    matches!(cx.qpath_res(&qpath, c.hir_id), Res::Def(DefKind::Ctor(..), ..))
+                },
+                ExprKind::AddrOf(.., expr) => helper(cx, expr),
+                ExprKind::Array(exprs) | ExprKind::Tup(exprs) => {
+                    for expr in exprs {
+                        if !helper(cx, expr) {
+                            return ControlFlow::Break(());
+                        }
+                    }
+
+                    true
+                },
+                ExprKind::Struct(_, fields, None) => {
+                    for field in fields {
+                        if !helper(cx, field.expr) {
+                            return ControlFlow::Break(());
+                        }
+                    }
+
+                    true
+                },
+                ExprKind::Path(qpath) => {
+                    // Can't compare a local with another local in a pat
+                    !matches!(cx.qpath_res(&qpath, expr.hir_id), Res::Local(..))
```

Comment:
Looks like you were right @blyxyas

We should probably do a positive `matches!` here: statics, const generics, associated consts and maybe others wouldn't be valid to suggest converting to a pattern

There's also a more subtle one, constants can only be used in a pattern if they implement [`StructuralEq`](https://doc.rust-lang.org/std/marker/trait.StructuralEq.html) and [`StructuralPartialEq`](https://doc.rust-lang.org/std/marker/trait.StructuralPartialEq.html):

```rust
struct S {
    a: usize,
}

impl PartialEq for S {
    fn eq(&self, _: &Self) -> bool {
        true
    }
}

impl Eq for S {}

static CONST_S: S = S { a: 1 };

fn f(opt_s: Option<S>) {
    match opt_s {
        Some(x) if x == CONST_S => {},
        // Some(CONST_S) would be a compile error
        _ => {},
    }
}

Reply:
> positive matches! here

This would be even worse, as you cannot compare a local with another local (and would also skip *any* structs or enums, therefore making this lint entirely pointless).

Though good catch

I think instead we should only allow stuff like `DefKind::Struct`, `DefKind::Enum`, etc, on a case-by-case basis

Reply:
Yeah that's what I meant by positive, `matches!(.., stuff we allow)` not `!matches(.., stuff we forbid)`